packages/playground/blueprints/src/lib/zip.ts

```diff
@@ -0,0 +1,170 @@
+import { FileEntry, readAllBytes } from '@php-wasm/universal';
+
+export interface ZipFileHeader {
+	startsAt?: number;
+	extract?: any;
+	signature: string;
+	version: number;
+	generalPurpose: number;
+	compressionMethod: number;
+	lastModifiedTime: number;
+	lastModifiedDate: number;
+	crc: number;
+	compressedSize: number;
+	uncompressedSize: number;
+	fileNameLength: number;
+	fileName: string;
+	extraLength: number;
+	extra: string;
+}
+
+export interface ZipFileEntry extends ZipFileHeader {
+	dataStream: ReadableStream<Uint8Array>;
+}
+
+const SIGNATURE_FILE = 0x04034b50;
+const SIGNATURE_CENTRAL_DIRECTORY_START = 0x02014b50;
+const SIGNATURE_CENTRAL_DIRECTORY_END = 0x06054b50;
+
+async function* iterateZipEntries(stream: ReadableStream<Uint8Array>) {
+	const zipStream = new BufferedReadableStream(stream.getReader());
+	while (true) {
+		const signature = await zipStream.readUint32();
+		// console.log({ signature });
+		if (signature === SIGNATURE_FILE) {
+			const header = await readFileHeader(zipStream);
+			const fileDataStream = await createFileDataStream(
+				zipStream,
+				header
+			);
+			yield {
+				...header,
+				dataStream: fileDataStream,
+			} as ZipFileEntry;
+		} else if (signature === SIGNATURE_CENTRAL_DIRECTORY_START) {
+			break;
+		} else if (signature === SIGNATURE_CENTRAL_DIRECTORY_END) {
+			break;
+		} else {
+			throw new Error(`Unknown signature: ${signature}`);
+		}
+	}
+}
+
+export async function* iterateZipFiles(stream: ReadableStream<Uint8Array>) {
+	for await (const entry of iterateZipEntries(stream)) {
+		yield toZipFile(entry);
+	}
+}
+
+function toZipFile(entry: ZipFileEntry): FileEntry {
+	return {
+		path: entry.fileName,
+		isDirectory: entry.fileName.endsWith('/'),
+		async read() {
+			return await readAllBytes(entry.dataStream.getReader());
+		},
+	};
+}
+
+async function readFileHeader(
+	stream: BufferedReadableStream
+): Promise<ZipFileHeader> {
+	const entry: Partial<ZipFileEntry> = {};
+	entry['version'] = await stream.readUint16();
+	entry['generalPurpose'] = await stream.readUint16();
+	entry['compressionMethod'] = await stream.readUint16();
+	entry['lastModifiedTime'] = await stream.readUint16();
+	entry['lastModifiedDate'] = await stream.readUint16();
+	entry['crc'] = await stream.readUint32();
+	entry['compressedSize'] = await stream.readUint32();
+	entry['uncompressedSize'] = await stream.readUint32();
+	entry['fileNameLength'] = await stream.readUint16();
+	entry['extraLength'] = await stream.readUint16();
+	entry['fileName'] = new TextDecoder().decode(
```

Comment:
it appears that Emscripten is the real culprit here, but filenames are not encoded. they are bytes separated by ASCII `/`. this will end up breaking filenames and inserting `�` in places.